### PR TITLE
Refactor Psim and Support Sensor Mocking

### PR DESCRIPTION
**Pull request by shihaocao**
# Refactor Psim and Support Sensor Mocking

Fixes #. (Optional; sometimes PRs don't have an associated ticket)

### Summary of changes
- Obliterates inheritance for simulation (everything is just CppSimulation)
- Obliterates "interact_fc", should be more transparent calling
- Adds member variable `self.mock_sensor_validity` of a CppSimulation, that gates whether or not ptest will attempt to mock piksi, adcs_havt, ssa_mode per #687 

### Base changes
- Includes enum print changes
- Includes changes from Fatima's configuration for HITL/HOOTL so that scraping uplinks is defined by the config files

### Testing
Tested in dual_psim, and psim_debug, and seems to be mocking away

### Constants
None

### Telemetry
None

### Documentation Evidence
This PR and inline documentation will be sufficient

### CODE REVIEW

File: ptest/cases/utils.py [hunk 1]
```diff
@@ -191,6 +191,12 @@ class Enums(object):
         "MTR_ENABLED"
     ])
 
+    ssa_modes = FSWEnum([
+        "SSA_FAILURE",
+        "SSA_IN_PROGRESS",
+        "SSA_COMPLETE",
+    ])
+
```
Comment (shihaocao) [inline ptest/cases/utils.py:199]: additional enums just nice to have for printing from PTest terminal

File: ptest/cases/utils.py [hunk 2]
```diff
@@ -221,7 +227,8 @@ def __getitem__(self, key):
             "adcs.state" : self.adcs_states,
             "radio.state" : self.radio_states,
             "sat.designation" : self.sat_designations,
-            "piksi.mode" : self.piksi_modes,
+            "piksi.state" : self.piksi_modes,
```
Comment (shihaocao) [inline ptest/cases/utils.py:230]: piksi.mode was never the name of the FSW state field

File: ptest/psim/simulation.py [hunk 3]
```diff
@@ -13,14 +13,15 @@
 import psim
 import lin 
 import json
-from ..cases.utils import is_lin_vector, to_lin_vector
+from ..cases.utils import is_lin_vector, to_lin_vector,Enums
+import math
 
-class Simulation(object):
+class CppSimulation(object):
```
Comment (shihaocao) [inline ptest/psim/simulation.py:19]: Renaming the class is the simplest way to go

File: ptest/psim/simulation.py [hunk 4]
```diff
@@ -306,22 +223,6 @@ def write_adcs_estimator_inputs(self, fc):
...
-        """
-
-        flight_controller.read_state("pan.state")
-        flight_controller.read_state("adcs.state")        
-        flight_controller.read_state("adcs_monitor.mag1_vec")
-        flight_controller.read_state("adcs_monitor.mag2_vec")
-        flight_controller.read_state("attitude_estimator.q_body_eci")
-        flight_controller.read_state("attitude_estimator.w_body")
-        flight_controller.read_state("attitude_estimator.fro_P")
-
```
Comment (shihaocao) [inline ptest/psim/simulation.py:324]: Extraneous. Logging is handled by the cases themselves

File: ptest/psim/simulation.py [hunk 5]
```diff
@@ -330,9 +231,109 @@ def update_dynamics(self):
...
+            # Step 2. Load sensor data from psim into ptest
+            self.update_sensors()
+                        
+
+            # print(self.mock_sensor_validity)
+            if self.mock_sensor_validity:
+                for device_name, device in self.devices.items():
+                    self.mock_piksi_state(device_name, device)
+                    self.mock_adcs_havt(device_name, device)
+                    self.mock_ssa_mode(device_name, device)
```
Comment (shihaocao) [inline ptest/psim/simulation.py:296]: actual new section

File: ptest/psim/simulation.py (hunk 3 above)
Comment (kylekrol) [reply ptest/psim/simulation.py:19]: What exactly was the naming collusion here?

Comment (shihaocao) [reply ptest/psim/simulation.py:19]: Ptest dependencies

File: ptest/runsim.py [hunk 6]
```diff
@@ -97,7 +95,8 @@ def set_up_devices(self):
                     # pty isn't defined because we're on Windows
                     self.stop_all(f"Cannot connect to a native binary for device {device_name}, since the current OS is Windows.")
 
-            device_session = USBSession(device_name, self.uplink_console, device["http_port"], is_teensy, self.simulation_run_dir, self.tlm_config, device['imei'], self.scrape_uplinks)
+            device_session = USBSession(device_name, self.uplink_console, device["http_port"], is_teensy, self.simulation_run_dir, 
+                self.tlm_config, device['imei'], device["scrape_uplinks"])
```
Comment (shihaocao) [inline ptest/runsim.py:99]: scrape uplinks read from config file and passed down

File: ptest/psim/simulation.py (hunk 3 above)
Comment (shihaocao) [reply ptest/psim/simulation.py:19]: Im just lazy, trust this is the way xd

File: ptest/psim/simulation.py [hunk 7]
```diff
@@ -295,6 +170,48 @@ def update_sensors(self):
                     psim_val = list(psim_val)
                 self.sensor_readings[role][fc_sf] = psim_val
 
+    def mock_piksi_state(self, fc_name, fc_device):
```
Comment (shihaocao) [inline ptest/psim/simulation.py:173]: new method that is all encompassing

File: ptest/psim/simulation.py [hunk 8]
```diff
@@ -295,6 +170,48 @@ def update_sensors(self):
...
+        fsw_piksi_state = -1
+
+        if cdgps_active == 0:
+            fsw_piksi_state = Enums.piksi_modes['spp']
+        else:
+            fsw_piksi_state = Enums.piksi_modes['fixed_rtk']
+
+        fc_device.write_state('piksi.state', fsw_piksi_state)
+
+    def mock_adcs_havt(self, fc_name, fc_device):
```
Comment (shihaocao) [inline ptest/psim/simulation.py:190]: new method that is all encompassing

File: ptest/psim/simulation.py [hunk 9]
```diff
@@ -295,6 +170,48 @@ def update_sensors(self):
...
+        '''
+        Set the gyro, and both mags to be "working"
+        '''
+        fc_device.write_state('adcs_monitor.havt_device0', True)
+        fc_device.write_state('adcs_monitor.havt_device1', True)
+        fc_device.write_state('adcs_monitor.havt_device2', True)
+
+    def mock_ssa_mode(self, fc_name, fc_device):
```
Comment (shihaocao) [inline ptest/psim/simulation.py:198]: new method that is all encompassing

File: ptest/psim/simulation.py [hunk 10]
```diff
@@ -330,9 +231,108 @@ def update_dynamics(self):
...
+                if type(local) == list:
+                    _len = len(local)
+                    if _len == 2:
+                        local = lin.Vector2(local)
+                    elif _len == 3:
+                        local = lin.Vector3(local)
+                    elif _len == 4:
+                        local = lin.Vector4(local)
+                    else:
+                        raise RuntimeError("Unexpected List Length, can't change into lin Vector")
```
Comment (shihaocao) [inline ptest/psim/simulation.py:266]: not new, just moved it

File: ptest/psim/simulation.py [hunk 11]
```diff
@@ -330,9 +231,108 @@ def update_dynamics(self):
...
+
+            # Step 2. Load sensor data from psim into ptest
+            self.update_sensors()
+                        
+            # Step 3.1 Mock sensor validity flags and states if requested
+            if self.mock_sensor_validity:
+                for device_name, device in self.devices.items():
+                    self.mock_piksi_state(device_name, device)
+                    self.mock_adcs_havt(device_name, device)
+                    self.mock_ssa_mode(device_name, device)
```
Comment (shihaocao) [inline ptest/psim/simulation.py:295]: new calling section for sensor mocking. Everything before this is not new

File: ptest/psim/simulation.py [hunk 12]
```diff
@@ -354,23 +354,13 @@ def read_actuators(self, fc):
...
-                        local = lin.Vector4(local)
-                    else:
-                        raise RuntimeError("Unexpected List Length, can't change into lin Vector")
+    def stop(self, data_dir):
+        """
+        Stops a run of the simulation and saves run data to disk.
+        """
+        self.add_to_log("Stopping simulation...")
+        self.running = False
+        time.sleep(1) # Wait for logs to finish for the current timestep.
```
Comment (shihaocao) [inline ptest/psim/simulation.py:363]: not new

File: ptest/psim/simulation.py [hunk 13]
```diff
@@ -354,23 +354,13 @@ def read_actuators(self, fc):
...
 
-                self.mysim[psim_sf] = local
+        with open(data_dir + "/simulation_log.txt", "w") as fp:
+            fp.write(self.log)
```
Comment (shihaocao) [inline ptest/psim/simulation.py:366]: not new

File: ptest/uplink_console.py [hunk 14]
```diff
@@ -45,8 +45,7 @@ def create_uplink(self, fields, vals, filename, json_filename):
 
         # Create a JSON file with all the fields and values
         telem_json={}
-        for field, val in zip(fields, vals):
-            value = self.get_val(val)
+        for field, value in zip(fields, vals):
```
Comment (shihaocao) [inline ptest/uplink_console.py:48]: @fatimayousuf what are these changes for?

File: ptest/usb_session.py [hunk 15]
```diff
@@ -482,7 +482,7 @@ def scrape_uplink(self):
             for response_part in data:
                 if isinstance(response_part, tuple):
                     # converts message from byte literal to string removing b''
-                    msg = email.message_from_string(response_part[1].decode('utf-8'))
+                    msg = email.message_from_bytes(response_part[1])
```
Comment (shihaocao) [inline ptest/usb_session.py:485]: @fatimayousuf I assume this is critical and now correct

File: src/fsw/FCCode/FieldCreatorTask.hpp [hunk 16]
```diff
@@ -37,7 +40,8 @@ class FieldCreatorTask : public ControlTask<void> {
 
         uplink_pos_f("orbit.uplink_pos", Serializer<lin::Vector3d>(6771000, 6921000, 28)),
         uplink_vel_f("orbit.uplink_vel", Serializer<lin::Vector3d>(7570, 7685, 19)),
-        uplink_t_f("orbit.uplink_t", Serializer<double>(0.0, 18'446'744'073'709'551'616.0, 64))
+        uplink_t_f("orbit.uplink_t", Serializer<double>(0.0, 18'446'744'073'709'551'616.0, 64)),
+        uplink_gps_t("uplink.time", Serializer<gps_time_t>())
```
Comment (shihaocao) [inline src/fsw/FCCode/FieldCreatorTask.hpp:44]: is this correct? @kylekrol

**Review by Duncan-McD — APPROVED** [review]
looks epicness

**Pull request merged**